File: api/src/api/schemas/catalog_migration.py
"""Request/response shapes for catalog storage-backend migration."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class MigrationStartRequest(BaseModel):
    target_storage_backend_id: uuid.UUID


class CatalogMigrationTableOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    schema_name: str
    table_name: str
    status: str
    bytes_copied: int
    error: str | None = None


class CatalogMigrationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    catalog_id: uuid.UUID
    source_storage_backend_id: uuid.UUID
    target_storage_backend_id: uuid.UUID
    status: str
    tables_total: int
    tables_done: int
    bytes_total: int
    bytes_copied: int
    error: str | None = None
    created_at: datetime
    updated_at: datetime
    started_at: datetime | None = None
    cutover_at: datetime | None = None
    finished_at: datetime | None = None
    # Populated on the detail endpoint; omitted from the list view.
    tables: list[CatalogMigrationTableOut] | None = None


class CatalogMigrationEventOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    seq: int
    level: str
    message: str
    created_at: datetime

File: web/src/types/catalog-migration.ts
// A run of moving a catalog's Iceberg data from one storage backend to another.
// Mirrors the API `CatalogMigrationOut` (api/schemas/catalog_migration.py).
export type MigrationStatus =
  | "pending"
  | "copying"
  | "verifying"
  | "cutover"
  | "completed"
  | "failed"
  | "cancelled";

export const MIGRATION_TERMINAL: MigrationStatus[] = [
  "completed",
  "failed",
  "cancelled",
];

export interface CatalogMigrationTable {
  schema_name: string;
  table_name: string;
  status: string;
  bytes_copied: number;
  error: string | null;
}

export interface CatalogMigration {
  id: string;
  catalog_id: string;
  source_storage_backend_id: string;
  target_storage_backend_id: string;
  status: MigrationStatus;
  tables_total: number;
  tables_done: number;
  bytes_total: number;
  bytes_copied: number;
  error: string | null;
  created_at: string;
  updated_at: string;
  started_at: string | null;
  cutover_at: string | null;
  finished_at: string | null;
  tables?: CatalogMigrationTable[] | null;
}

export interface CatalogMigrationEvent {
  seq: number;
  level: string;
  message: string;
  created_at: string;
}
